In [1]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [2]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [3]:
# Generate summary statistics
df_market_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 41 entries, bitcoin to digibyte
Data columns (total 7 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   price_change_percentage_24h   41 non-null     float64
 1   price_change_percentage_7d    41 non-null     float64
 2   price_change_percentage_14d   41 non-null     float64
 3   price_change_percentage_30d   41 non-null     float64
 4   price_change_percentage_60d   41 non-null     float64
 5   price_change_percentage_200d  41 non-null     float64
 6   price_change_percentage_1y    41 non-null     float64
dtypes: float64(7)
memory usage: 2.6+ KB


In [4]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [5]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
market_data_scaled = StandardScaler().fit_transform(df_market_data[df_market_data.columns])

In [6]:
# Create a DataFrame with the scaled data
df_market_data_scaled = pd.DataFrame(market_data_scaled,
                                     columns=df_market_data.columns)

# Copy the crypto names from the original DataFrame
df_market_data_scaled['coin_id'] = df_market_data.index

# Set the coin_id column as index
df_market_data_scaled.set_index('coin_id', inplace=True)

# Display the scaled DataFrame
df_market_data_scaled

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [7]:
# Create a list with the number of k-values from 1 to 11
k = range(1, 11)

In [8]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list

for i in k:
    k_model = KMeans(n_clusters=i, random_state=1)
    k_model.fit(df_market_data_scaled)
    inertia.append(k_model.inertia_)


c:\Users\jitra\anaconda3\envs\dev\lib\site-packages\joblib\externals\loky\backend\context.py:136: UserWarning: Could not find the number of physical cores for the following reason:
[WinError 2] The system cannot find the file specified
Returning the number of logical cores instead. You can silence this warning by setting LOKY_MAX_CPU_COUNT to the number of cores you want to use.
  warnings.warn(
  File "c:\Users\jitra\anaconda3\envs\dev\lib\site-packages\joblib\externals\loky\backend\context.py", line 257, in _count_physical_cores
    cpu_info = subprocess.run(
  File "c:\Users\jitra\anaconda3\envs\dev\lib\subprocess.py", line 503, in run
    with Popen(*popenargs, **kwargs) as process:
  File "c:\Users\jitra\anaconda3\envs\dev\lib\subprocess.py", line 971, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "c:\Users\jitra\anaconda3\envs\dev\lib\subprocess.py", line 1456, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
c

In [9]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {
    "k" : k,
    "inertia" : inertia
}

# Create a DataFrame with the data to plot the Elbow curve
elbow_df = pd.DataFrame(elbow_data)

In [10]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
elbow_df.hvplot.line(
    x = "k",
    y = "inertia",
    title = "Elbow Plot"
)

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** Four (4)

---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [11]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=3, random_state=1)

In [12]:
# Fit the K-Means model using the scaled DataFrame
model.fit(df_market_data_scaled)

c:\Users\jitra\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=3, random_state=1)

In [13]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
predictons = model.predict(df_market_data_scaled)

# Print the resulting array of cluster values.
print(predictons)

[2 2 0 0 2 2 2 2 2 0 0 0 0 2 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 0 0 0 0 2 0 0 1
 0 0 0 0]


In [14]:
# Create a copy of the scaled DataFrame
scaled_market_data_df = df_market_data_scaled.copy()

In [15]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
scaled_market_data_df['cluster'] = predictons

# Display the copy of the scaled DataFrame
scaled_market_data_df

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

In [16]:
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.

scaled_market_data_df.hvplot.scatter(
    x = "price_change_percentage_24h",
    y = "price_change_percentage_7d",
    color = "cluster",
    hover_cols = 'coin_id'
)


:Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,cluster,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [17]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [18]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components.
scaled_market_data_pca = pca.fit_transform(scaled_market_data_df)

# View the scaled PCA data
scaled_market_data_pca[:5]

array([[ 1.53989072e+00,  1.72074200e-03,  5.99429843e-01],
       [ 1.26430978e+00, -2.03956315e-02,  1.15828613e+00],
       [-3.11943499e-01, -5.56328469e-01, -7.93148023e-01],
       [-3.08756016e-01, -6.07235705e-01, -6.43243271e-01],
       [ 2.97256712e+00,  1.18690730e-01,  1.28739950e+00]])

In [19]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
display(pca.explained_variance_ratio_)
display(f"The total explained variance of the three principal components is: {pca.explained_variance_ratio_.sum():.3%}")


array([0.35447676, 0.32842379, 0.20180846])

'The total explained variance of the three principal components is: 88.471%'

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** As shown above, just below 88.5% which is pretty good!

In [20]:
# Create a new DataFrame with the PCA data.
df_market_data_pca = pd.DataFrame(scaled_market_data_pca)

# Copy the crypto names from the original scaled DataFrame
crypto_names = df_market_data.index

# Set the coin_id column as index
df_market_data_pca.set_index(crypto_names, inplace=True)

# Display the scaled PCA DataFrame
df_market_data_pca

0         1         2
coin_id                                           
bitcoin               1.539891  0.001721  0.599430
ethereum              1.264310 -0.020396  1.158286
tether               -0.311943 -0.556328 -0.793148
ripple               -0.308756 -0.607236 -0.643243
bitcoin-cash          2.972567  0.118691  1.287400
binancecoin           1.987592  0.339342  0.697284
chainlink             1.642718  0.108206  2.711209
cardano               1.520289  0.228621  1.513142
litecoin              1.265414 -0.219908  0.853886
bitcoin-cash-sv      -0.113912 -0.838170 -0.453656
crypto-com-chain     -1.482896 -0.978600 -1.031230
usd-coin             -0.319061 -0.565413 -0.809163
eos                  -0.413812 -0.919334 -0.614740
monero                3.038145  1.571615  0.788478
tron                 -0.172684 -0.462779 -0.290292
tezos                 0.695514 -0.743133  1.572005
okb                  -1.475700 -0.635592 -0.739420
stellar              -0.623362 -0.831081 -0.115052
cosmos                1.524004  0.263562  1.721661
cdai                 -0.259100 -0.617068 -0.822398
neo                  -0.996064 -0.866577 -0.565775
wrapped-bitcoin       1.525498 -0.009536  0.585877
leo-token            -0.789257 -0.787963 -0.967236
huobi-token          -0.306617 -0.539076 -0.533986
nem                  -0.303430  0.670457 -1.388138
binance-usd          -0.291935 -0.564097 -0.805477
iota                 -0.281837 -0.964302  0.012186
vechain              -1.740466 -1.493170 -0.680064
zcash                 1.295731  0.029996  1.237759
theta-token          -1.820172  2.203501 -1.663179
dash                 -0.298384 -0.807303  0.118214
ethereum-classic     -0.296907 -0.744190 -0.297121
ethlend              -6.544986  5.457553  3.696990
maker                 0.907373 -0.118967  0.800828
havven               -2.360560 -0.301951  0.835899
omisego              -0.187647  0.168050 -1.415395
celsius-degree-token  2.904469  7.231151 -3.478616
ontology             -1.736001 -1.607926 -0.130986
ftx-token             0.058777 -0.569863 -0.166715
true-usd             -0.274309 -0.563614 -0.801521
digibyte             -0.432496 -0.458893 -0.983983

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [21]:
# Create a list with the number of k-values from 1 to 11
k_pca = list(range(1, 11))

In [22]:
# Create an empty list to store the inertia values
inertia_pca = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list

for i in k_pca:
    k_model_pca = KMeans(n_clusters=i, random_state=1)
    k_model_pca.fit(df_market_data_pca)
    inertia_pca.append(k_model_pca.inertia_)





c:\Users\jitra\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\jitra\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\jitra\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\jitra\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is k

In [23]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_pca = {
    "k_pca" : k_pca,
    "inertia_pca" : inertia_pca
}
# Create a DataFrame with the data to plot the Elbow curve
elbow_pca_df = pd.DataFrame(elbow_data_pca)

elbow_pca_df

k_pca  inertia_pca
0      1   285.070506
1      2   198.049575
2      3   118.645628
3      4    43.556391
4      5    33.156756
5      6    24.137539
6      7    19.392111
7      8    16.496539
8      9    12.947640
9     10     9.938473

In [24]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
elbow_pca_df.hvplot.line(
    x = "k_pca",
    y = "inertia_pca",
    title = "Elbow Plot"
)



:Curve   [k_pca]   (inertia_pca)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** Four, as this is where the elbow curve kinks / starts to flatten.


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** It does not differ based on the answe above.

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [25]:
# Initialize the K-Means model using the best value for k
K_model_pca2 = KMeans(n_clusters=4, random_state=1)

In [26]:
# Fit the K-Means model using the PCA data
K_model_pca2.fit(df_market_data_pca)

c:\Users\jitra\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=1)

In [27]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
predictions_pca = K_model_pca2.predict(df_market_data_pca)

# Print the resulting array of cluster values.
print(predictions_pca)

[1 1 0 0 1 1 1 1 1 0 0 0 0 1 0 1 0 0 1 0 0 1 0 0 0 0 0 0 1 0 0 0 3 1 0 0 2
 0 0 0 0]


In [28]:
# Create a copy of the scaled PCA DataFrame
scaled_pca_data = df_market_data_pca.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
scaled_pca_data["cluster"] = predictions_pca

# Display the copy of the scaled PCA DataFrame
scaled_pca_data

0         1         2  cluster
coin_id                                                    
bitcoin               1.539891  0.001721  0.599430        1
ethereum              1.264310 -0.020396  1.158286        1
tether               -0.311943 -0.556328 -0.793148        0
ripple               -0.308756 -0.607236 -0.643243        0
bitcoin-cash          2.972567  0.118691  1.287400        1
binancecoin           1.987592  0.339342  0.697284        1
chainlink             1.642718  0.108206  2.711209        1
cardano               1.520289  0.228621  1.513142        1
litecoin              1.265414 -0.219908  0.853886        1
bitcoin-cash-sv      -0.113912 -0.838170 -0.453656        0
crypto-com-chain     -1.482896 -0.978600 -1.031230        0
usd-coin             -0.319061 -0.565413 -0.809163        0
eos                  -0.413812 -0.919334 -0.614740        0
monero                3.038145  1.571615  0.788478        1
tron                 -0.172684 -0.462779 -0.290292        0
tezos                 0.695514 -0.743133  1.572005        1
okb                  -1.475700 -0.635592 -0.739420        0
stellar              -0.623362 -0.831081 -0.115052        0
cosmos                1.524004  0.263562  1.721661        1
cdai                 -0.259100 -0.617068 -0.822398        0
neo                  -0.996064 -0.866577 -0.565775        0
wrapped-bitcoin       1.525498 -0.009536  0.585877        1
leo-token            -0.789257 -0.787963 -0.967236        0
huobi-token          -0.306617 -0.539076 -0.533986        0
nem                  -0.303430  0.670457 -1.388138        0
binance-usd          -0.291935 -0.564097 -0.805477        0
iota                 -0.281837 -0.964302  0.012186        0
vechain              -1.740466 -1.493170 -0.680064        0
zcash                 1.295731  0.029996  1.237759        1
theta-token          -1.820172  2.203501 -1.663179        0
dash                 -0.298384 -0.807303  0.118214        0
ethereum-classic     -0.296907 -0.744190 -0.297121        0
ethlend              -6.544986  5.457553  3.696990        3
maker                 0.907373 -0.118967  0.800828        1
havven               -2.360560 -0.301951  0.835899        0
omisego              -0.187647  0.168050 -1.415395        0
celsius-degree-token  2.904469  7.231151 -3.478616        2
ontology             -1.736001 -1.607926 -0.130986        0
ftx-token             0.058777 -0.569863 -0.166715        0
true-usd             -0.274309 -0.563614 -0.801521        0
digibyte             -0.432496 -0.458893 -0.983983        0

In [29]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.

# Rename column with integer index 1 to 'PCA1'
scaled_pca_data.rename(columns={1: 'PCA1', 2:'PCA2'}, inplace=True)

# Create the scatter plot with hvplot
scaled_pca_data.hvplot.scatter(
    x='PCA1',
    y='PCA2', 
    color='cluster',
    hover_cols=['coin_id']
)

:Scatter   [PCA1]   (PCA2,cluster,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [30]:
# Composite plot to contrast the Elbow curves
pca_elbow = elbow_pca_df.hvplot.line(
    x = "k_pca",
    y = "inertia_pca",
    title = "Composite Elbow Plot",
    color = "orange",
    label = "PCA"
)

elbow = elbow_df.hvplot.line(
    x = "k",
    y = "inertia",
    title = "Composite Elbow Plot",
    color = "blue",
    label = "Non-optimized"
)

composite_elbow = pca_elbow + elbow

composite_elbow


:Layout
   .Curve.PCA                        :Curve   [k_pca]   (inertia_pca)
   .Curve.Non_hyphen_minus_optimized :Curve   [k]   (inertia)

In [31]:
# Composite plot to contrast the clusters
pca_data = scaled_pca_data.hvplot.scatter(
    x = 'PCA1',
    y = 'PCA2',  
    color = 'cluster',
    cmap = 'Viridis',
    hover_cols = 'coin_id',
    label = "PCA",
    marker = "o",
    colorbar = True
).opts(colorbar_position = 'right')

data = scaled_market_data_df.hvplot.scatter(
    x = "price_change_percentage_24h",
    y = "price_change_percentage_7d",
    color = "cluster",
    cmap = "Plasma",
    hover_cols = 'coin_id',
    label = "Non-optimized",
    marker = "^",
    colorbar = True
).opts(colorbar_position = "right")

composite_data = pca_data + data
composite_data

:Layout
   .Scatter.PCA                        :Scatter   [PCA1]   (PCA2,cluster,coin_id)
   .Scatter.Non_hyphen_minus_optimized :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,cluster,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** While the optimal number of K is unchanged, by using PCA, we were able to reduce the scatter. The process resulted in keeping the data columns with the most predictive power. We can still see two outliers in the PCA model however. If we zoom into the PCA data, we can see some postive relationship between both PCA variables between -2 and 1 (albeit having some fanning around the scatter as we move up PCA1 values). This is not so true in the non-optimized model, where it is hard to see any real positive relationship between price_change variables (e.g. cannot infer that 24hr price changes have a positive relationship with 7d changes). 